### PR TITLE
Spi

### PR DESCRIPTION
**Pull request by sp-bechara**

### CODE REVIEW

File: Task/MCU/Floor1_USART_inputdata/Core/Src/main.c [hunk 1]
```diff
@@ -141,14 +151,10 @@ int adcInterruptCheckFlag=0;
...
 #ifdef SPI
-uint8_t rx_buffer[2] = {0, 0};
-uint8_t tx_buffer[2] = {0xD7, 0};
-uint8_t DeviceIDRxBuffer[8] = {0, 0, 0, 0, 0, 0, 0, 0};
-uint8_t DeviceIDTxBuffer[8] = {0x9F, 0, 0, 0, 0, 0, 0};
-// Data to be transferred
-uint8_t data[256];
-uint8_t readRx[261];
-uint8_t memoryPageReadRx[264];
+//uint8_t readRx[261];
```
Comment (kaushaleinfo) [inline Task/MCU/Floor1_USART_inputdata/Core/Src/main.c]: Put this inside debug_SPI_Flash macro

File: Task/MCU/Floor1_USART_inputdata/Core/Src/main.c [hunk 2]
```diff
@@ -234,6 +240,103 @@ void watchdogFeed(void){
...
+
+int deviceReady(){
+	while(1){
+		uint8_t rx_buffer[3] = {0, 0, 0};
+		uint8_t tx_buffer[3] = {0xD7};
+		spiChipSelect();
+		HAL_SPI_TransmitReceive(&hspi1, tx_buffer, rx_buffer, 3, 1000);
+		spiChipDeselect();
+		if(((rx_buffer[1]>>BIT_NUMBER) & 1)==1){
+			HAL_UART_Transmit(&huart2, (uint8_t *)"Device is ready\n", sizeof("Device is ready\n"), 1000);
```
Comment (kaushaleinfo) [inline Task/MCU/Floor1_USART_inputdata/Core/Src/main.c:274]: Put this inside debug_SPI_Flash macro

File: Task/MCU/Floor1_USART_inputdata/Core/Src/main.c [hunk 3]
```diff
@@ -234,6 +240,103 @@ void watchdogFeed(void){
...
+			return 1;
+			break;
```
Comment (kaushaleinfo) [inline Task/MCU/Floor1_USART_inputdata/Core/Src/main.c:276]: Put this break appropriate space

File: Task/MCU/Floor1_USART_inputdata/Core/Src/main.c [hunk 4]
```diff
@@ -234,6 +240,103 @@ void watchdogFeed(void){
...
+		}
+	}
+}
+
+void memoryWrite(uint32_t address, uint8_t* buffer, int noOfBytes)
+{
```
Comment (kaushaleinfo) [inline Task/MCU/Floor1_USART_inputdata/Core/Src/main.c:282]: Put vaild address check

File: Task/MCU/Floor1_USART_inputdata/Core/Src/main.c [hunk 5]
```diff
@@ -234,6 +240,103 @@ void watchdogFeed(void){
...
+
+        spiChipSelect();
+        HAL_SPI_Transmit(&hspi1, commandTx, sizeof(commandTx), HAL_MAX_DELAY);
+        HAL_SPI_Transmit(&hspi1, buffer, (noOfBytes > 256) ? 256 : noOfBytes, HAL_MAX_DELAY);
+        spiChipDeselect();
+
+        //Condition to check whether device is ready or not
+        deviceReady();
+
+        //Read data from buffer
```
Comment (kaushaleinfo) [inline Task/MCU/Floor1_USART_inputdata/Core/Src/main.c:304]: Put this inside debug_SPI_Flash macro

File: Task/MCU/Floor1_USART_inputdata/Core/Src/main.c [hunk 6]
```diff
@@ -234,6 +240,103 @@ void watchdogFeed(void){
...
+        HAL_SPI_Transmit(&hspi1, commandTx, sizeof(commandTx), 1000);
+        spiChipDeselect();
+
+        // Update the variables
+        noOfBytes -= PAGE_SIZE;
+        buffer += PAGE_SIZE;
+        address += PAGE_SIZE;
+
+        //Condition to check whether device is ready or not
+        deviceReady();
```
Comment (kaushaleinfo) [inline Task/MCU/Floor1_USART_inputdata/Core/Src/main.c]: Put this function at appropriate location

File: Task/MCU/Floor1_USART_inputdata/Core/Src/main.c [hunk 7]
```diff
@@ -234,6 +240,103 @@ void watchdogFeed(void){
...
+    uint8_t memoryPageReadTx[262] = {0x1B, address >> 16, (address >> (POSITION_OF_BITS - 1)) & ((1 << NO_OF_BITS_TO_BE_EXTRACTED) - 1), (address >> (POSITION_OF_BITS_FOR_BYTES_ADDRESS - 1)) & ((1 << NO_OF_BITS_TO_BE_EXTRACTED_FOR_BYTES_ADDRESS) - 1), 0x00, 0x00};
+
+    while (readRangeCheck > 0) {
+        spiChipSelect();
+        memoryPageReadTx[1] = address >> 16;
+        memoryPageReadTx[2] = (address>>(POSITION_OF_BITS-1)) & ((1<<NO_OF_BITS_TO_BE_EXTRACTED)-1);
+        memoryPageReadTx[3] = (address >> (POSITION_OF_BITS_FOR_BYTES_ADDRESS - 1)) & ((1 << NO_OF_BITS_TO_BE_EXTRACTED_FOR_BYTES_ADDRESS) - 1);
+        HAL_SPI_TransmitReceive(&hspi1, memoryPageReadTx, memoryPageReadRx, PAGE_SIZE+6, 1000);
+        spiChipDeselect();
+        readRangeCheck=(readRange >= 256) ? 261 :readRange+7;
```
Comment (kaushaleinfo) [inline Task/MCU/Floor1_USART_inputdata/Core/Src/main.c]: what is 256 and 261 ?

**Review by kaushaleinfo — CHANGES_REQUESTED** [review]

File: Task/MCU/Floor1_USART_inputdata/Core/Src/main.c (hunk 1 above)
Comment (sp-bechara) [reply Task/MCU/Floor1_USART_inputdata/Core/Src/main.c]: done

File: Task/MCU/Floor1_USART_inputdata/Core/Src/main.c (hunk 2 above)
Comment (sp-bechara) [reply Task/MCU/Floor1_USART_inputdata/Core/Src/main.c:274]: done

File: Task/MCU/Floor1_USART_inputdata/Core/Src/main.c (hunk 3 above)
Comment (sp-bechara) [reply Task/MCU/Floor1_USART_inputdata/Core/Src/main.c:276]: done

File: Task/MCU/Floor1_USART_inputdata/Core/Src/main.c (hunk 5 above)
Comment (sp-bechara) [reply Task/MCU/Floor1_USART_inputdata/Core/Src/main.c:304]: As discussed, we don't need to do it anymore...As only print inside it is only kept under debug macro.

File: Task/MCU/Floor1_USART_inputdata/Core/Src/main.c (hunk 6 above)
Comment (sp-bechara) [reply Task/MCU/Floor1_USART_inputdata/Core/Src/main.c]: done

File: Task/MCU/Floor1_USART_inputdata/Core/Src/main.c (hunk 7 above)
Comment (sp-bechara) [reply Task/MCU/Floor1_USART_inputdata/Core/Src/main.c]: done

File: Task/MCU/Floor1_USART_inputdata/Core/Src/main.c (hunk 4 above)
Comment (sp-bechara) [reply Task/MCU/Floor1_USART_inputdata/Core/Src/main.c:282]: done

**Review by kaushaleinfo — APPROVED** [review]

**Pull request merged**